blog/content/post/2021-10-28-secure-jtag-on-imx-soc.md

```diff
@@ -0,0 +1,270 @@
+
+Would you really like to blow this fuse? Type YES all uppercase to confirm: 
+YES
+soc:IMX6UL ref:1 otp:JTAG_SMODE op:blow addr:0x18 off:22 len:2 val:0x1 res:0x00004000
+
+
+```
+OpenOCD with i.MX secure SJC patch use `imx6_sjcauth.txt` file to store access
+keys. Challenge key we can get during the first try of connection. Let's try:
+
+```
+user@machine:~/openocd-code/tcl$ sudo openocd -f interface/ftdi/olimex-arm-usb-ocd-h.cfg -f target/imx6.cfg
+Open On-Chip Debugger 0.9.0-dev-00019-g970a12aef-dirty (2021-10-11-14:19)
+Licensed under GNU GPL v2
+For bug reports, read
+    http://openocd.sourceforge.net/doc/doxygen/bugs.html
+Info : only one transport option; autoselect 'jtag'
+Warn : imx6.sdma: nonstandard IR value
+adapter speed: 10 kHz
+Info : clock speed 10 kHz
+SJC  : Using database tcl/target/imx6_sjcauth.txt
+SJC  : Challenge 0x0e81d4eacb3785
+SJC  : Response  0x574c20308fad77
+SJC  : Using database tcl/target/imx6_sjcauth.txt
+SJC  : Challenge 0x0e81d4eacb3785
+SJC  : Response  0x574c20308fad77
+SJC  : Using database tcl/target/imx6_sjcauth.txt
+SJC  : Challenge 0x0e81d4eacb3785
+SJC  : Response  0x574c20308fad77
+Info : JTAG tap: imx6.dap tap/device found: 0x4ba00477 (mfg: 0x23b, part: 0xba00
+, ver: 0x4)
+Info : TAP imx6.sdma does not have IDCODE
+Info : JTAG tap: imx6.sjc tap/device found: 0x1891a01d (mfg: 0x00e, part: 0x891a
+, ver: 0x1)
+Info : imx6.cpu.0: hardware has 6 breakpoints, 4 watchpoints
+Info : number of cache level 1
+Info : imx6.cpu.0 cluster 0 core 0 multi core
+```
+
+Works fine, now a response key is required to access JTAG. Remember to lock mode
+and key overriding by burning relevant fuses.
+
+Different case, different device - suppose that I finish development and this
+product never be debugged. Now I permanently disable this possibility:
+
+```
+# ./crucible -m IMX6UL -r1 -b 16 -e big blow JTAG_SMODE 0x3
+```
+
+```
+# ./crucible -m IMX6UL -r 1 -b 16 -e big blow BOOT_CFG_LOCK 0x3
+```
+
+When I try to get access by JTAG port I can see that i.MX doesn't allow for
+that:
+
+```
+user@machine:~$ sudo openocd -f interface/ftdi/olimex-arm-usb-ocd-h.cfg -f target/imx6.cfg
+Open On-Chip Debugger 0.11.0
+Licensed under GNU GPL v2
+For bug reports, read
+    http://openocd.org/doc/doxygen/bugs.html
+Info : auto-selecting first available session transport "jtag". To override use 'transport select <transport>'.
+Warn : imx6.sdma: nonstandard IR value
+Info : Listening on port 6666 for tcl connections
+Info : Listening on port 4444 for telnet connections
+Info : clock speed 10 kHz
+Error: JTAG scan chain interrogation failed: all zeroes
+Error: Check JTAG interface, timings, target power, etc.
+Error: Trying to use configured scan chain anyway...
+Error: imx6.cpu: IR capture error; saw 0x00 not 0x01
+Warn : Bypassing JTAG setup events due to errors
+Error: Invalid ACK (0) in DAP response
+```
+
+### Using U-boot
+
+## Summary
+
+Secure JTAG mode is a very useful feature, which should be implemented in every
```

Comment:
Let's explain why leaving JTAG on the board may be useful and why exactly securing it is something we should do. Try to come up with some examples.